File: python-algorithm/algo_02_sliding_window/L567-m.py
# -*- coding: utf-8 -*-


# 滑动窗口

# 1 收缩窗口 判断条件：排序后是否相等，速度比较慢
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        lt, rt = 0, 0
        goal = self.sort_string(s1)
        while rt < len(s2):
            while rt-lt+1 == len(s1):
                if self.sort_string(s2[lt:rt+1]) == goal:
                    return True
                lt += 1
            rt += 1
        return False

    def sort_string(self, s):
        return "".join((lambda x: (x.sort(), x)[1])(list(s)))


# 2 收缩窗口的判断条件，巧妙的设计了一个变量cnt（76题也可以这么做），来判断子串是不是s1的排列
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        # record用于记录窗口中可用的s1中的字符个数，有点绕
        record = {}
        for c in s1:
            if c not in record:
                record[c] = 1
            else:
                record[c] += 1
        lt, rt = 0, 0
        cnt = 0 # 标记窗口中出现s1中的字符的个数
        while rt < len(s2):
            if s2[rt] in s1:
                record[s2[rt]] -= 1
                if record[s2[rt]] >= 0:
                    cnt += 1
            # 收缩窗口左边界
            while rt-lt+1 == len(s1):
                if cnt == len(s1):
                    return True
                if s2[lt] in s1:
                    record[s2[lt]] += 1
                    if record[s2[lt]] >= 1:
                        cnt -= 1
                lt += 1
            rt += 1
        return False

File: python-algorithm/algo_02_sliding_window/L567-m.cpp

#include <iostream>
#include <cstring>
#include <vector>
using namespace std ;


//首先这道题用暴力解法超时

//因此用的是滑动窗口解法
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        int sz1 = s1.size();
        int sz2 = s2.size();
        if (sz1 > sz2) return false;
        vector<int> v1(128, 0);
        vector<int> v2(128, 0);
        for(auto c:s1)
            ++v1[c];
        int l=0;
        for(int r=0; r<sz2; ++r){
            ++v2[s2[r]];
            if(r-l+1 == sz1){
                if (v1 == v2)
                    return true;
                --v2[s2[l]];
                ++l;
            }
        }
        return false;
    }
};


//这种解法是从L76转化过来，不同的是，更新窗口的时机不一样
//L76是当前字符出现了
class Solution1 {
public:
    bool checkInclusion(string s1, string s2) {
        vector<int> chars(128, 0);
        vector<bool> flag(128, false);
        // 先统计T中的字符情况
        for(int i = 0; i < s1.size(); ++i) {
            flag[s1[i]] = true;
            ++chars[s1[i]]; //为了处理s1中重复字母
        }
        // 移动滑动窗口，不断更改统计数据
        int cnt = 0, l = 0;
        for (int r = 0; r < s2.size(); ++r) {
            if (flag[s2[r]]) {
                if (--chars[s2[r]] >= 0)
                    ++cnt;
            }
            // 若目前滑动窗口长度等于s1长度，判断是否是s1的全排列
            if(r-l+1 == s1.size()){
                if (cnt == s1.size())
                    return true;
                if (flag[s2[l]] && ++chars[s2[l]] > 0)
                    --cnt;
                ++l;
            }
        }
        return false;
    }
};

int main(){
    string s1 = "ab";
    string s2 = "eidboaoo";
    cout << Solution().checkInclusion(s1,s2)<<endl;
}